File: fourSum.cpp
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>

using namespace std;

//#define testMod

bool cmp(int a,int b)
{
    return a < b; //升序
}

#ifdef testMod
void test()
{
    
}
#endif

#ifndef testMod

class Solution {
private:


public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> out;
        if(nums.size() < 4)
            return out;
        sort(begin(nums),end(nums),cmp);
        int high,low,mid,want;
        for(unsigned long long i=0;i<nums.size()-3;i++)
        {
            if(i!=0&&nums[i]==nums[i-1])
                continue;
            for(unsigned long long j=i+1;j<nums.size()-2;j++)
            {
                if(j!=i+1&&nums[j]==nums[j-1])
                    continue;
                for(unsigned long long k=j+1;k<nums.size()-1;k++)
                {
                    if(k!=j+1&&nums[k]==nums[k-1])
                        continue;
                    cout<<i<<j<<k<<endl;
                    low=k+1;
                    high=nums.size()-1;
                    want=target-nums[i]-nums[j]-nums[k];

                    while(low<=high)
                    {
                        mid=(high+low)/2;
                        cout<<i<<j<<k<<high<<low<<mid<<endl;
                        if(nums[mid]==want)
                        {
                            out.push_back({nums[i],nums[j],nums[k],nums[mid]});
                            break;
                        }
                        else if(nums[mid]<want)
                        {
                            low=mid+1;
                        }
                        else
                        {
                            high=mid-1;
                        }
                    }
                }
            }
        }
        return out;
    }
};
#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;
    vector<int> nums ={1, 0, -1, 0, -2, 2};
    int target = 0;
    vector<vector<int>> out = sl.fourSum(nums,target);
    for(unsigned long long i=0;i<out.size();i++)
    {
        for(unsigned long long j=0;j<out[1].size();j++)
        {
            cout<<out[i][j]<<" ";
        }
        cout<<endl;
    }

#endif 
    return 0;
}

File: fourSum.py
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()
        n = len(nums)

        for i in range(n - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, n - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                t = target - nums[i] - nums[j]
                h = n - 1

                for k in range(j + 1, n - 1):
                    if k > j + 1 and nums[k] == nums[k - 1]:
                        continue

                    while k < h and nums[k] + nums[h] > t:
                        h -= 1

                    if k >= h:
                        break

                    if nums[k] + nums[h] == t:
                        res.append([nums[i], nums[j], nums[k], nums[h]])

        return res


nums = [1, 0, -1, 0, -2, 2]
target = 0
nums = [2, 2, 2, 2, 2]
target = 8
print(Solution().fourSum(nums, target))
